research/picai/nnunet_scripts/eval.py

```diff
@@ -1,26 +1,188 @@
+    case_identifiers: Optional[Iterable[str]] = None,
+    **kwargs: Any
+) -> PicaiEvalMetrics:
+    """
+    Computes the picai evaluation metrics provided the predicted lesion
+    detection maps and the ground truth annotations. Extends picai_eval to
+    allow multiclass evaluation
+
+    Args:
+        detection_maps (NDArray): The predicted lesion detection maps. Must
+            have shape (num_samples, num_lesion_classes, ...). The background
+            class should not be included in the detection maps
+        ground_truth_annotations (NDArray): The ground truth annotations. Must
+            have shape (num_samples, num_classes or num_lesion_classes, ...).
+            If num_classes is provided, the function will attempt to remove
+            the background class from index 0 for you
+        case_identifiers (Optional[Iterable[str]], optional): A list of case
+            identifiers. If not provided the subjects will be identified by
+            their index Defaults to None.
+        **kwargs: Keyword arguments for the picai_eval.evaluate function
+
+    Raises:
+        KeyError: If you try to use the y_det, y_true, or subject_list keyword
+            arguments
+
+    Returns:
+        picai_eval.metrics.Metrics: A picai eval metrics object that has
+            combined the results from all classes into a single object
+    """
+    if "y_det" in kwargs or "y_true" in kwargs or "subject_list" in kwargs:
+        raise KeyError(
+            """Got one of 'y_det', 'y_true' or 'subject_list' in keyword
+            arguments. The arguments to get_picai_metrics are being passed to
+            these keys"""
+        )
+
+    if detection_maps.shape[1] != ground_truth_annotations.shape[1]:
+        # Assume background class has not been removed
+        ground_truth_annotations = ground_truth_annotations[:, 1:]
+        assert (
+            detection_maps.shape == ground_truth_annotations.shape
+        ), "Got unexpected shapes for detection maps and ground truth annotations"
+
+    # Evaluation must be calculated seperately for each class
+    num_classes = detection_maps.shape[1]
+    metrics: List[PicaiEvalMetrics] = []
+    for cls in range(num_classes):
+        metrics.append(
+            evaluate(
+                y_det=detection_maps[:, cls],
+                y_true=ground_truth_annotations[:, cls],
+                subject_list=case_identifiers,
+                **kwargs
+            )
+        )
+
+    subject_list = metrics[0].subject_list
+    assert isinstance(subject_list, list), "Got unexpected subject list from picai eval metrics object"
+    lesion_results: Dict[Any, list] = defaultdict(list)
+    lesion_weights: Dict[Any, list] = defaultdict(list)
+    case_targets: Dict[Any, int] = defaultdict(int)
+    case_preds: Dict[Any, float] = defaultdict(float)
+    for s in subject_list:
+        # Ignoring mypy errors here for now because i don't know how to get around them
```

Comment:
```python
for s in subject_list:
  for m in metrics:
        assert (
            m.lesion_results is not None
            and not isinstance(m.lesion_results, Path)
            and not isinstance(m.lesion_results, os.PathLike)
            and not isinstance(m.lesion_results, str)
        )
        lesion_results[s].extend(m.lesion_results[s])
        assert m.lesion_weight is not None
        lesion_weights[s].extend(m.lesion_weight[s])

    # Should probably do an assertion that none of the m.case_target/m.case_pred are None instead of filtering 
    case_targets[s] = int(any(m.case_target[s] for m in metrics if m.case_target is not None))
    case_preds[s] = max([m.case_pred[s] for m in metrics if m.case_pred is not None])
```

Reply:
I was able to avoid this mess by hooking into the lower level function. I was trying to avoid doing that before, but with the new version that uses disk, there was no way around without giving up support for multiclass segmentation